### PR TITLE
[Data] [5/11] Add the Parquet FooterReader actor pool

### PR DESCRIPTION
**Pull request by goutamvenkat-anyscale**
## Why

Footer IO for the footer-based Parquet chunking path (the rest of the split of #64985).

Reading N Parquet footers serially on the driver doesn't scale — it's the step that decides which row groups a read even touches. `FooterReader` runs as a pool of Ray actors, each reading footers on its own thread pool and streaming results back in batches, so footer IO spreads across the cluster instead of bottlenecking on one process.

**Nothing constructs it yet** — it's unused until the footer indexer lands in step 6.

## Flow

Per file, it produces the row-group descriptors the bin packer (merged in #65210) consumes:

```mermaid
flowchart LR
  B["(path, size) batch"] --> A["FooterReaderActor<br/>SPREAD across the cluster"]
  A --> R[read footer]
  R --> F["split_by_row_group(filter)<br/>prune by statistics"]
  F --> Q["split_by_row_group(~filter)<br/>classify fully_matched"]
  Q --> C["coalesce_row_groups<br/>(from #65210)"]
  C --> O["stream List[FileChunks]<br/>batched by result_batch_size"]
```

Three things it decides, each of which the packer and the limit push-down depend on:

- **Predicate pruning.** `fragment.split_by_row_group(filter)` drops row groups whose Parquet statistics contradict the pushed filter, so they never reach the packer at all.
- **`fully_matched` classification.** A surviving group is *fully* matched iff `~filter` prunes it entirely. Only fully-matched groups have an exact survivor count, and only exact counts may drive limit push-down. Any failure defaults to not-fully-matched, which is always the safe direction. **This is the correctness-critical bit**: nulls are the interesting case, since Parquet min/max stats are computed over non-null values only, so a group whose non-null values all satisfy the filter looks fully matched by bounds alone while its null rows don't survive.
- **Projection-scoped sizing.** Byte sizes are accounted over only the projected leaf columns, so bins get sized for what the reader will actually fetch. Nested columns expand to all their leaves.

## What changed

Two new files, 368 insertions, 0 deletions:

| File | Change |
| --- | --- |
| `listing/footer_reader.py` | **new** — `FooterReader` plus `FooterReaderActor = ray.remote(FooterReader)`. Kept as a plain class with the actor built via the functional form, so callers can type handles as `ActorProxy[FooterReader]`. `read_footers` is a streaming `@ray.method` generator that yields `List[FileChunks]`, batched by `result_batch_size` — the driver pays one object-store fetch per yielded list, so batching cuts driver-side deserialization roughly proportionally. |
| `tests/test_footer_reader.py` | **new** — 14 tests |

## Testing

The tests drive the plain class directly against local Parquet fixtures — no actor, no cluster. Coverage is aimed at the decisions above rather than the plumbing: predicate pruning by statistics, the partial-match case (`id >= 30` splitting a row group, which must **not** count as exact), a predicate matching nothing, projection shrinking accounted size while leaving row counts alone, coalescing preserving total rows, nested columns expanding to all leaves, and result batching at several `result_batch_size` values.

```
$ python -m pytest -q python/ray/data/_internal/datasource_v2/tests/test_footer_reader.py
14 passed in 4.30s

$ python -m pytest -q python/ray/data/_internal/datasource_v2/tests/ python/ray/data/tests/unit/datasource_v2/
145 passed in 0.90s
```

## Stack

Step 5 of an 11-step split of #64985, which was 43 files / +3234 −719 and not reviewable as one unit.

- #65167 `[1/11]` Extract `FileIndexer.list_file_infos` ✅ merged
- #65168 `[2/11]` Push limit into `ReadFiles` when `Limit` sits directly on it ✅ merged
- #65214 `[3/11]` Derive `ListFiles` pushdown state from the scanner ✅ merged
- #65210 `[4/11]` Parquet footer types + online bin packer ✅ merged
- **this PR** `[5/11]` — depends on #65210 for `parquet_footer_types` and `parquet_row_group_coalescing`
- #65169 `[9/11]` Pin the Parquet footer actor pool to 1 in tests (open, independent)

Step 6 wires this and the packer together behind an opt-in flag, and needs both this and #65214.

---
Not a duplicate of any open PR (checked `gh pr list --state open` plus a file-level overlap check against #65142 / #63158 / #64404 — no overlap; only #64985, which this replaces).
AI assistance was used; every line reviewed by me and tests run locally.

### CODE REVIEW

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+            # out. For those, num_rows is an exact survivor count and can drive
+            # the limit push-down. Any failure defaults to "not fully matched",
+            # which is always safe (falls back to the post-filter stop).
+            try:
+                not_fully = {
+                    sub.row_groups[0].id
+                    for sub in fragment.split_by_row_group(~self.filter)
+                }
+            except Exception:
+                not_fully = set(rg_indices)
```
[the remark this answers is not in the document]
Comment (goutamvenkat-anyscale) [reply python/ray/data/_internal/datasource_v2/listing/footer_reader.py]: Not an issue in pyarrow 23

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        # Projection pushdown: when set, row-group byte sizes are accounted over
+        # only these top-level columns, since the reader will only fetch those.
+        self.projected_cols: Optional[Set[str]] = (
+            set(projected_cols) if projected_cols is not None else None
+        )
+        # Reused across files; make_fragment is what lets us apply
+        # split_by_row_group.
+        self.file_format = pds.ParquetFileFormat()
+
+    def _projected_leaf_indices(self, row_group) -> Optional[List[int]]:
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py]:
```suggestion
    def _projected_leaf_indices(self, row_group: RowGroupMetaData) -> Optional[List[int]]:
```

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+        # Map the requested top-level column names to Parquet leaf-column indices
+        # (a nested field expands to several leaves, e.g. "a.b.list.element").
+        # Returns None to signal "all columns" so callers can take the cheap path.
+        if self.projected_cols is None:
+            return None
+        return [
+            j
+            for j in range(row_group.num_columns)
+            if _leaf_matches(row_group.column(j).path_in_schema, self.projected_cols)
+        ]
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py]: In hive-style partitioning, a column's value can be encoded in the directory name instead of inside the files (e.g. `s3://bucket/events/year=2024/part-0.parquet`). I think this might lead to a peculiar edge case where `self.projected_cols` only contain synthetic columns (`year` in the above example) here. The function would return `[]` in that case, leading to `_row_group_uncompressed_size(rg, [])` returning 0. This would end up putting all row groups in bin (since `used_bytes` would always stay 0), and consequently in one task, breaking parallelism. 

_Schema drift might also lead to this issue but is much less likely since we sample the schema across 16 files._

I believe the following would patch the issue:
```suggestion
        indices = [
            j
            for j in range(row_group.num_columns)
            if _leaf_matches(row_group.column(j).path_in_schema, self.projected_cols)
        ]
        return indices or None
```

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+    def _row_group_info(
+        self,
+        row_group: RowGroupMetaData,
+        rg_idx: int,
+        leaf_indices: Optional[List[int]],
+        fully_matched: bool = True,
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py]: Although this only has one call site, I feel `fully_matched` should default to `False` just to be safe, since having a false positive for `fully_matched` would lead to queries silently returning fewer rows than asked for.

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+        # the V1 helper so the three call sites cannot drift.
+        uncompressed = _row_group_uncompressed_size(row_group, leaf_indices)
+        return RowGroupInfo(
+            rg_idx=rg_idx,
+            uncompressed_size=uncompressed,
+            num_rows=row_group.num_rows,
+            fully_matched=fully_matched,
+        )
+
+    def _has_filter_nulls(self, metadata, rg_idx: int) -> bool:
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py]:
```suggestion
    def _has_filter_nulls(self, metadata: FileMetaData, rg_idx: int) -> bool:
```

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+        matched: Set[str] = set()
+        for j in range(row_group.num_columns):
+            column = row_group.column(j)
+            if not _leaf_matches(column.path_in_schema, self.filter_columns):
+                continue
+            matched.update(
+                name
+                for name in self.filter_columns
+                if column.path_in_schema == name
+                or column.path_in_schema.startswith(f"{name}.")
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py]: Nit: I feel we can extract the predicate used here as a helper function since the same predicate is used in `_leaf_matches` and line 99 to 102 in `parquet_file_reader.py`:

```python
def _prefix_match(path: str, names: Set[str]) -> Set[str]:
    return {n for n in names if path == n or path.startswith(f"{n}.")}
```
Then `_leaf_matches` is just:
```python
def _leaf_matches(leaf_path: str, names: Set[str]) -> bool:
    return bool(_prefix_match(leaf_path, names))
```

Also the `_leaf_matches` and `matched.updated` actually repeated the same linear computation for the predicate.

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+            )
+            stats = column.statistics
+            if stats is None or not stats.has_null_count or stats.null_count > 0:
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py:222]: Nit: I think we don't yet have tests that capture the cases of `not stats.has_null_count` or `stats.null_count > 0`.

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+        group's ``num_rows`` is not an exact survivor count.
+
+        Missing or null-count-less statistics are treated as "may contain
+        nulls", i.e. not fully matched -- the safe direction.
+        """
+        row_group = metadata.row_group(rg_idx)
+        matched: Set[str] = set()
+        for j in range(row_group.num_columns):
+            column = row_group.column(j)
+            if not _leaf_matches(column.path_in_schema, self.filter_columns):
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py]: I think we can get all the leaf-matching indices in `_read_and_chunk` and pass it into `_has_filter_nulls` instead of checking this every call, considering each Parquet file only has one schema. We can cut the runtime of this logic from O(M*N) to O(M+N) in that case. Would look like how `leaf_indices` is computed on line 210.

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+            # ~filter cannot match any of its rows, i.e. the negation prunes it
+            # out. For those, num_rows is an exact survivor count and can drive
+            # the limit push-down. Any failure defaults to "not fully matched",
+            # which is always safe (falls back to the post-filter stop).
+            try:
+                not_fully = {
+                    sub.row_groups[0].id
+                    for sub in fragment.split_by_row_group(~self.filter)
+                }
+            except Exception:
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py]: Nit: Do we need to log the Exception and some context here, or is it okay as it is given that partial match is a safe case?

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+            self._projected_leaf_indices(metadata.row_group(0))
+            if metadata.num_row_groups
+            else None
+        )
+        per_rg = [
+            self._row_group_info(
+                metadata.row_group(i),
+                i,
+                leaf_indices,
+                fully_matched=(True if fully_by_idx is None else fully_by_idx[i]),
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py]:
```suggestion
                fully_matched=fully_by_idx.get(i, True),
```

File: python/ray/data/_internal/datasource_v2/tests/test_footer_reader.py [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+        path, size = _write(
+            tmp_path / "n.parquet",
+            pa.table({"id": pa.array(values, pa.int64())}),
+            row_group_size=len(values),
+        )
+
+        chunks = _reader(filter_expr=col("id") >= 30)._read_and_chunk(path, size)
+
+        claimed = sum(rg.num_rows for rg in chunks.row_groups if rg.fully_matched)
+        assert claimed <= expected_survivors
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/tests/test_footer_reader.py]: This check feels a bit lenient, might be better to directly assert `claimed == expected_claimed`?

File: python/ray/data/_internal/datasource_v2/tests/test_footer_reader.py [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+
+        # The reader only fetches projected columns, so bin sizing must account
+        # for those bytes alone -- otherwise bins are sized for data never read.
+        assert sum(rg.uncompressed_size for rg in projected.row_groups) < sum(
+            rg.uncompressed_size for rg in full.row_groups
+        )
+        # Row counts are a property of the file, not the projection.
+        assert [rg.num_rows for rg in projected.row_groups] == [
+            rg.num_rows for rg in full.row_groups
+        ]
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/tests/test_footer_reader.py]: Thinking the test could be more rigorous if it's something like this:
```suggestion
def test_projection_accounts_only_the_projected_columns(self, four_row_groups):
    path, size = four_row_groups

    full = _reader()._read_and_chunk(path, size)
    ids  = _reader(projected_cols=["id"])._read_and_chunk(path, size)
    pads = _reader(projected_cols=["pad"])._read_and_chunk(path, size)

    total = lambda c: sum(rg.uncompressed_size for rg in c.row_groups)
    assert total(ids) + total(pads) == total(full)
    assert 0 < total(ids) < total(full)
    assert [rg.num_rows for rg in ids.row_groups] == [rg.num_rows for rg in full.row_groups]
```

**Review by rayhhome — CHANGES_REQUESTED** [review]
Sorry for the late comments!

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py (hunk 7 above)
Comment (goutamvenkat-anyscale) [reply python/ray/data/_internal/datasource_v2/listing/footer_reader.py:222]: Added `test_file_without_statistics_fails_closed`

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py (hunk 8 above)
Comment (goutamvenkat-anyscale) [reply python/ray/data/_internal/datasource_v2/listing/footer_reader.py]: Good catch. We can cache per file.

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py (hunk 9 above)
Comment (goutamvenkat-anyscale) [reply python/ray/data/_internal/datasource_v2/listing/footer_reader.py]: Added a debug log for the exception message.

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+        # Returns None to signal "all columns" so callers can take the cheap path.
+        if self.projected_cols is None:
+            return None
+        indices = [
+            j
+            for j in range(row_group.num_columns)
+            if _leaf_matches(row_group.column(j).path_in_schema, self.projected_cols)
+        ]
+
+        return indices or None
```
[the remark this answers is not in the document]
Comment (goutamvenkat-anyscale) [reply python/ray/data/_internal/datasource_v2/listing/footer_reader.py:160]: This comment is not applicable since if the column doesn't match any of the prefixes, that error should be caught when doing the projection pushdown not when doing the listfiles operation.

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+            uncompressed_size=uncompressed,
+            num_rows=row_group.num_rows,
+            fully_matched=fully_matched,
+        )
+
+    def _locate_filter_columns(self, row_group: RowGroupMetaData) -> _FilterLeaves:
+        # Resolve the predicate's column names to Parquet leaf-column indices,
+        # the same mapping ``_projected_leaf_indices`` does for the projection.
+        # Called once per file so the per-row-group null check below is an
+        # index lookup instead of another scan over every leaf.
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py]: Let's do the same thing as `_projected_leaf_indices` and have an early return for `_locate_filter_columns` if `filter_columns` is `None`:
```python
        if not self.filter_columns:
            return _FilterLeaves([], True)
```

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 15]
```diff
@@ -0,0 +1,313 @@
...
+        results stream out as footers land.
+        """
+        futures = [
+            self.pool.submit(self._read_and_chunk, path, size) for path, size in files
+        ]
+        buffer: list[FileChunks] = []
+        total_row_groups = 0
+        for finished in as_completed(futures):
+            chunk = finished.result()
+            total_row_groups += len(chunk.row_groups)
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py:391]: If we're not doing any per file bookkeeping here, we can skip a vacuous yield and the accompanying `ray.get` down the line:
```python
            if not chunk.row_groups:
                # Every row group was pruned
                continue
```

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 16]
```diff
@@ -0,0 +1,313 @@
...
+        callers set it via ``RAY_DATA_PARQUET_FOOTER_RESULT_BATCH_SIZE``.
+        ``num_returns`` is fixed to ``"streaming"`` via ``@ray.method`` so
+        results stream out as footers land.
+        """
+        futures = [
+            self.pool.submit(self._read_and_chunk, path, size) for path, size in files
+        ]
+        buffer: list[FileChunks] = []
+        total_row_groups = 0
+        for finished in as_completed(futures):
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py]: Does the `FooterReader` need to honor the `preserve_order` execution option? I think `as_completed` here returns the futures in the order they finish, which is indeterministic and can mean different manifest blocks generated for downstream work.

However, while switching to `for finished in futures:` would address the determinism issue, yielding in the exact order means we would face the edge case where a heavy file comes first in the queue. In this scenario, we would have to wait for the heavy futures to complete first, before getting to the ones that finished earlier but are at the back of the queue.

So my idea for this is:
```python
        ordered_futures = futures if preserve_order else as_completed(futures)
        for finished in ordered_futures:
```

Not sure if this is overkill though.

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 17]
```diff
@@ -0,0 +1,313 @@
...
+        row_group = metadata.row_group(rg_idx)
+        for j in filter_leaves.indices:
+            stats = row_group.column(j).statistics
+            if stats is None or not stats.has_null_count or stats.null_count > 0:
+                return True
+        return False
+
+    def _read_and_chunk(self, path: str, size: int) -> FileChunks:
+        fragment = self.file_format.make_fragment(path, filesystem=self.filesystem)
+        metadata = fragment.metadata  # reads + caches the footer on the fragment
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py]: This line may raise exception when the file read is truncated, non-parquet, corrupted, or deleted right after listing, so we might need a try-except guard here. Also need to decide whether to ignore the file or abort the process altogether.

File: python/ray/data/_internal/datasource_v2/tests/test_footer_reader.py [hunk 18]
```diff
@@ -0,0 +1,401 @@
...
+
+        # Measured rather than restated: this guard exists because a null row
+        # passes neither the filter nor its negation, so pin what the filter
+        # actually returns instead of trusting the number above.
+        assert pq.read_table(path, filters=[("id", ">=", 30)]).num_rows == survivors
+
+        exact_rows = sum(rg.num_rows for rg in chunks.row_groups if rg.fully_matched)
+        assert exact_rows == expected_exact_rows
+
+    def test_null_free_group_is_still_exact(self, tmp_path):
```
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/tests/test_footer_reader.py]: I think what this test examines has been roughly covered by the `no-nulls-is-exact` parametrization in the `test_exact_counts_come_only_from_null_free_groups` test above.

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py (hunk 13 above)
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py:160]: I think we can add a test that examines the case where `_projected_leaf_indices` returns `None`.

File: python/ray/data/_internal/datasource_v2/readers/parquet_file_reader.py
Comment (rayhhome) [inline python/ray/data/_internal/datasource_v2/readers/parquet_file_reader.py:116]: We can simplify the logic here with `_row_group_uncompressed_size`:
```
    indices = None
    if columns is not None:
        projected = set(columns)
        indices = [
            j
            for j in range(row_group_meta.num_columns)
            if _leaf_matches(row_group_meta.column(j).path_in_schema, projected)
        ]
    row_group_uncompressed_size = _row_group_uncompressed_size(row_group_meta, indices)
```

**Review by rayhhome — COMMENTED** [review]
Some more comments!

**Comment by ayushk7102** [comment]
> ## What changed
> Two new files, 368 insertions, 0 deletions:

Nit: let's update the PR description

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 19]
```diff
@@ -0,0 +1,313 @@
...
+
+if TYPE_CHECKING:
+    import pyarrow.compute as pc
+
+    from ray.data.expressions import Expr
+
+logger = logging.getLogger(__name__)
+
+
+def _prefix_match(path: str, names: set[str]) -> set[str]:
```
Comment (ayushk7102) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py]:
```suggestion
def _get_prefix_matches(path: str, names: set[str]) -> set[str]:
```

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 20]
```diff
@@ -0,0 +1,313 @@
...
+    Kept as a plain class (the actor is created via the functional
+    ``ray.remote(FooterReader)`` form below) so callers can type actor handles as
+    ``ActorProxy[FooterReader]`` -- see Ray's type-hint docs
+    (https://docs.ray.io/en/latest/ray-core/type-hint.html).
+    """
+
+    def __init__(
+        self,
+        filesystem: pafs.FileSystem,
+        io_concurrency: int = 128,
```
Comment (ayushk7102) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py:95]: We should document why we arrive on 128 for a default `io_concurrency` value here. Maybe we should case this default on whether the filesystem is S3 or local?

i.e. something like:
```python
if (isinstance(filesystem, LocalFileSystem)):
    io_concurrency = 8 

elif (isinstance(filesystem, S3FileSystem)):
    io_concurrency = 128  # higher concurrency because network round-trip
```


High concurrency might make sense for remote reads from S3 (which I recognise is probably the most common case), but could this be a bottleneck when reading from HDD? The size of the I/O thread pool currently defaults to 8 threads which Arrow explicitly mentions in the docs as being a reasonable default for HDD

https://arrow.apache.org/docs/cpp/threading.html

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 21]
```diff
@@ -0,0 +1,313 @@
...
+class _FilterLeaves(NamedTuple):
+    """Where the predicate's columns sit in a file's schema.
+
+    A Parquet file has one schema, so this holds for all of its row groups and
+    is resolved once per file. ``all_columns_found`` is false when some
+    predicate column has no leaf in the schema at all.
+    """
+
+    indices: list[int]
+    all_columns_found: bool
```
Comment (ayushk7102) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py]: Nit: Can we rename 
```suggestion
    all_filter_columns_found: bool
```

existing name kind of sounds like all the parquet leaves are included in the predicate

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 22]
```diff
@@ -0,0 +1,313 @@
...
+            # test is currently safe by accident. Do not rely on that: check
+            # ``null_count`` directly, which holds regardless of how clever the
+            # engine's statistics pruning gets.
+            fully_by_idx: dict[int, bool] = {
+                i: i not in not_fully
+                and not self._has_filter_nulls(
+                    metadata=metadata, rg_idx=i, filter_leaves=filter_leaves
+                )
+                for i in rg_indices
+            }
```
Comment (ayushk7102) [inline python/ray/data/_internal/datasource_v2/listing/footer_reader.py:322]: Nit: This is kind of confusing because because `_has_filter_nulls()` can return True when the real reason is that the predicate column wasn't found. 

Can we break up this check into two parts inside a helper function named `_rg_can_fully_match()`? 
The conditions for the rg to be able to full match:
1) all predicate columns exist and 2) there are no null rows

```suggestion
             fully_by_idx: dict[int, bool] = {
                  i: i not in not_fully
                  and self._rg_can_fully_match(metadata.row_group(i), filter_leaves)
                  for i in rg_indices
              }

```


```
      def _has_filter_nulls(...)
          return ....

      def _rg_can_fully_match(
          self, row_group: RowGroupMetaData, filter_leaves: _FilterLeaves
      ) -> bool:
          """Whether this row group can fully contain matched rows.
         """
          if not filter_leaves.all_columns_found:
              return False
          return not self._has_filter_nulls(row_group, filter_leaves.indices) 

```

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py (hunk 17 above)
Comment (goutamvenkat-anyscale) [reply python/ray/data/_internal/datasource_v2/listing/footer_reader.py]: So non parquet files shouldn't go through this path. But we can guard with try-except

Comment (goutamvenkat-anyscale) [reply python/ray/data/_internal/datasource_v2/listing/footer_reader.py]: Actually I will wrap this call in a retry since it's IO bound and can fail

File: python/ray/data/_internal/datasource_v2/tests/test_footer_reader.py (hunk 18 above)
Comment (goutamvenkat-anyscale) [reply python/ray/data/_internal/datasource_v2/tests/test_footer_reader.py]: Removed

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py (hunk 13 above)
Comment (goutamvenkat-anyscale) [reply python/ray/data/_internal/datasource_v2/listing/footer_reader.py:160]: Added `test_none_when_projection_matches_no_leaf`

**Review by ayushk7102 — COMMENTED** [review]
Minor nits, otherwise LGTM

File: python/ray/data/_internal/datasource_v2/listing/footer_reader.py [hunk 23]
```diff
@@ -0,0 +1,313 @@
...
+        else:
+            leaf_indices = None
+            filter_leaves = _FilterLeaves([], False)
+
+        if self.filter is not None:
+            # Predicate pushdown: drop row groups whose Parquet statistics
+            # contradict the filter, so we never emit chunks the reader would
+            # skip anyway. Each returned fragment views a single surviving group.
+            surviving = fragment.split_by_row_group(self.filter)
+            rg_indices: Iterable[int] = [sub.row_groups[0].id for sub in surviving]
```
[the remark this answers is not in the document]
Comment (rayhhome) [reply python/ray/data/_internal/datasource_v2/listing/footer_reader.py]: I feel we probably need to handle this in some way; column evolution might introduce new columns in newly written parquet files. Then, if a filter on new columns get applied on old parquet files, it would raise an `ArrowInvalid` exception and crash. In the old path, reader uses `schema=self._file_dataset_schema` which is the unified schema, which gets around this issue.

**Review by rayhhome — APPROVED** [review]
LGTM; there's one potential edge case regression pointed out by cursor but i don't think it's blocking. Also check out Ayush's reviews!
